AX: Move roleDescription and rolePlatformDescription to AXCoreObject

hoffmanjoshua opened the pull request:
#### 0983be4b7d0cab884896063398694a7cdfdbf1d5
<pre>
AX: Move roleDescription and rolePlatformDescription to AXCoreObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=292692">https://bugs.webkit.org/show_bug.cgi?id=292692</a>
<a href="https://rdar.apple.com/150890881">rdar://150890881</a>

Reviewed by Tyler Wilcock.

Moving roleDescription and rolePlatformDescription to AXCoreObject allows
us to not have to cache the full role description for every single object
(when applicable). Instead, we can use some existing properties, as well
as a new one, InputType, to compute this on and off the main thread.

As part of this change, several properties that we cached previously can
be removed in favor of using the single InputType cached on inputs.

I&apos;ve also updated the WK1 test expectations for accessibility/roles-exposed.html.
The datetime-local, date, month, week, time are not supported on WK1, so we
shouldn&apos;t have those in our expectations.

* LayoutTests/accessibility/mac/input-type-role-description-dynamic-expected.txt: Added.
* LayoutTests/accessibility/mac/input-type-role-description-dynamic.html: Added.
* LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isFileUploadButton const):
(WebCore::AXCoreObject::isInputImage const):
(WebCore::AXCoreObject::supportsARIARoleDescription const):
(WebCore::AXCoreObject::isRadioInput const):
(WebCore::AXCoreObject::roleDescription):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleARIARoleDescriptionChanged):
(WebCore::AXObjectCache::handleInputTypeChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::handleRoleDescriptionChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityAttachment.cpp:
(WebCore::AccessibilityAttachment::attachmentElement const):
(WebCore::AccessibilityAttachment::roleDescription): Deleted.
* Source/WebCore/accessibility/AccessibilityAttachment.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isInputImage const): Deleted.
(WebCore::AccessibilityNodeObject::isRadioInput const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::inputType const):
(WebCore::AccessibilityObject::rolePlatformDescription): Deleted.
(WebCore::AccessibilityObject::supportsARIARoleDescription const): Deleted.
(WebCore::AccessibilityObject::roleDescription): Deleted.
(WebCore::AccessibilityObject::isFileUploadButton const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isMediaTimeline const):
(WebCore::AccessibilityObject::isSliderThumb const):
(WebCore::AccessibilityObject::isDescriptionList const): Deleted.
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::rolePlatformDescription): Deleted.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::inputType const):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: <a href="https://commits.webkit.org/294712@main">https://commits.webkit.org/294712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e286db7246e7ab54557bd4208c011c835404bab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107871 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53347 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104745 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30881 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78118 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35080 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105712 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92690 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58450 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17421 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10722 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52704 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87236 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10792 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110247 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29843 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22001 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87096 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86702 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22086 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31534 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9262 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24151 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29770 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35090 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29578 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -49,6 +49,14 @@ bool AXCoreObject::isList() const
+    if (std::optional type = inputType())
+        return *type == InputType::Type::File;
+
+    return false;
```
Another way to write this would be:

```cpp
std::optional type = inputType();
return type ? *type == InputType::Type::File : false;
```

hoffmanjoshua replied:
Updated this + the other places

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -75,6 +83,17 @@ bool AXCoreObject::isMenuItem() const
+    if (std::optional type = inputType())
+        return *type == InputType::Type::Image;
+
+    return false;
```
Same comment here.

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -805,6 +836,14 @@ bool AXCoreObject::isRootWebArea() const
+    if (std::optional type = inputType())
+        return *type == InputType::Type::Radio;
+
+    return false;
```
Same comment here.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2812,17 +2812,29 @@ void AXObjectCache::handleRoleChanged(AccessibilityObject& axObject, Accessibili
 
+void AXObjectCache::handleInputTypeChanged(Element& element)
+{
+    auto* object = get(element);
```
Should we store a `RefPtr` rather than raw pointer here, since `updateIsolatedTree` is not "trivial" per this documentation?

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack

hoffmanjoshua replied:
Yeah, that makes sense. I changed this location to a RefPtr, and also the other places where we store the result of `get(element)`, and use it non-trivially, in a raw pointer in this file.

twilco commented on Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
```diff
@@ -302,6 +302,118 @@ static void attributedStringSetStyle(NSMutableAttributedString *attributedString
+        return AXSummaryText();
+
+    return { };
+
```
Extra newline here.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -674,6 +677,10 @@ void AXIsolatedTree::updateNodeProperties(AccessibilityObject& axObject, const A
             properties.append({ AXProperty::IdentifierAttribute, axObject.identifierAttribute().isolatedCopy() });
             break;
+        case AXProperty::InputType:
+            if (std::optional inputType = axObject.inputType())
+                properties.append({ AXProperty::InputType, *inputType });
```
What if the `type` was changed to empty string? Would this mean the `std::optional` would be `std::nullopt`, and then we never change this property, so it remains its old non-empty value?

hoffmanjoshua replied:
`type` defaults to `Text`, even with an empty string, so we will still get a type here. See the implementation in `InputType::createIfDifferent`

twilco replied:
Ahh nice, thanks!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -645,6 +646,7 @@ void AXIsolatedObject::setProperty(AXProperty property, AXPropertyValueVariant&&
         [](RetainPtr<NSAttributedString>& typedValue) { return !typedValue; },
         [](RetainPtr<id>& typedValue) { return !typedValue; },
 #endif
+        [&](InputType::Type&) { return false; },
```
Why do we need the & capture here?

hoffmanjoshua replied:
We do not oops, removed that!

hoffmanjoshua committed:
AX: Move roleDescription and rolePlatformDescription to AXCoreObject
https://bugs.webkit.org/show_bug.cgi?id=292692
rdar://150890881

Reviewed by Tyler Wilcock.

Moving roleDescription and rolePlatformDescription to AXCoreObject allows
us to not have to cache the full role description for every single object
(when applicable). Instead, we can use some existing properties, as well
as a new one, InputType, to compute this on and off the main thread.

As part of this change, several properties that we cached previously can
be removed in favor of using the single InputType cached on inputs.

I've also updated the WK1 test expectations for accessibility/roles-exposed.html.
The datetime-local, date, month, week, time are not supported on WK1, so we
shouldn't have those in our expectations.

* LayoutTests/accessibility/mac/input-type-role-description-dynamic-expected.txt: Added.
* LayoutTests/accessibility/mac/input-type-role-description-dynamic.html: Added.
* LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isFileUploadButton const):
(WebCore::AXCoreObject::isInputImage const):
(WebCore::AXCoreObject::supportsARIARoleDescription const):
(WebCore::AXCoreObject::isRadioInput const):
(WebCore::AXCoreObject::roleDescription):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleARIARoleDescriptionChanged):
(WebCore::AXObjectCache::handleInputTypeChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::handleRoleDescriptionChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityAttachment.cpp:
(WebCore::AccessibilityAttachment::attachmentElement const):
(WebCore::AccessibilityAttachment::roleDescription): Deleted.
* Source/WebCore/accessibility/AccessibilityAttachment.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isInputImage const): Deleted.
(WebCore::AccessibilityNodeObject::isRadioInput const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::inputType const):
(WebCore::AccessibilityObject::rolePlatformDescription): Deleted.
(WebCore::AccessibilityObject::supportsARIARoleDescription const): Deleted.
(WebCore::AccessibilityObject::roleDescription): Deleted.
(WebCore::AccessibilityObject::isFileUploadButton const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isMediaTimeline const):
(WebCore::AccessibilityObject::isSliderThumb const):
(WebCore::AccessibilityObject::isDescriptionList const): Deleted.
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::rolePlatformDescription): Deleted.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::inputType const):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: https://commits.webkit.org/294712@main
